pkg/controller/tidbcluster/pod_control.go

```diff
@@ -0,0 +1,318 @@
+
+type stat struct {
+	observeAnnotationCounts int
+	finishAnnotationCounts  int
+}
+
+func (c *PodController) getPodStat(pod *corev1.Pod) stat {
+	c.podStatsMu.Lock()
+	defer c.podStatsMu.Unlock()
+
+	stat := c.podStats[pod.Namespace+pod.Name]
+	return stat
+}
+
+func (c *PodController) setPodStat(pod *corev1.Pod, stat stat) {
+	c.podStatsMu.Lock()
+	defer c.podStatsMu.Unlock()
+
+	c.podStats[pod.Namespace+pod.Name] = stat
+}
+
+// enqueueTidbCluster enqueues the given pod in the work queue.
+func (c *PodController) enqueuePod(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Cound't get key for object %+v: %v", obj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+// Run the controller.
+func (c *PodController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting tidbcluster pod controller")
+	defer klog.Info("Shutting down tidbcluster pod controller")
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+// worker runs a worker goroutine that invokes processNextWorkItem until the the controller's queue is closed
+func (c *PodController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem dequeues items, processes them, and marks them done. It enforces that the syncHandler is never
+// invoked concurrently with the same key.
+func (c *PodController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	result, err := c.sync(key.(string))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("TidbCluster pod: %v, sync failed %v, requeuing", key.(string), err))
+		c.queue.AddRateLimited(key)
+	} else {
+		if result.RequeueAfter > 0 {
+			c.queue.AddAfter(key, result.RequeueAfter)
+		} else if result.Requeue {
+			c.queue.AddRateLimited(key)
+		} else {
+			c.queue.Forget(key)
+		}
+	}
+	return true
+}
+
+func (c *PodController) sync(key string) (reconcile.Result, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	pod, err := c.deps.PodLister.Pods(ns).Get(name)
+	if errors.IsNotFound(err) {
+		klog.Infof("pod has been deleted %v", key)
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	pod = pod.DeepCopy()
+
+	// labels of tikv:
+	// app.kubernetes.io/component=tikv,app.kubernetes.io/instance=db1369682775200135879,app.kubernetes.io/managed-by=tidb-operator ...
+	managedBy := pod.Labels[label.ManagedByLabelKey]
+	if managedBy != "tidb-operator" {
+		return reconcile.Result{}, nil
+	}
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing TidbCluster pod %q (%v)", key, time.Since(startTime))
+	}()
+
+	component := pod.Labels[label.ComponentLabelKey]
+	ctx := context.Background()
+	switch component {
+	case "tikv":
```

Comment:
use `labels.IsTiKV` or `labels.TiKVLabelVal` directly? we are some consts and methods defined in `label.go`